### PR TITLE
Add comments to explain special cases

### PR DESCRIPTION
**Pull request by kerthcet**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:

~Based on my understanding, even `borrowlingLimit` not nil, if `cq.cohort == nil`, we should still not consider the borrowingLimit in calculation. cc @alculquicondor is this a real bug or as designed?~

Just for explaination.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
I agree that if the CQ doesn't belong to a cohort, the borrowingLimit should be considered 0.
Can you add a unit test case in TestPreemption that proves that there was a bug?

File: pkg/scheduler/preemption/preemption.go [hunk 1]
```diff
@@ -334,12 +336,15 @@ func workloadFits(wlReq cache.FlavorResourceQuantities, cq *cache.ClusterQueue,
 			}
 			for rName, rReq := range flvReq {
 				limit := flvQuotas.Resources[rName].Nominal
-				if flvQuotas.Resources[rName].BorrowingLimit != nil && allowBorrowing {
+
+				if cq.Cohort != nil && allowBorrowing && flvQuotas.Resources[rName].BorrowingLimit != nil {
```
Comment (kerthcet) [inline pkg/scheduler/preemption/preemption.go:340]: Based on the comment:
```
	// If null, it means that there is no borrowing limit.
	// If not null, it must be non-negative.
	// borrowingLimit must be null if spec.cohort is empty.
	// +optional
	BorrowingLimit *resource.Quantity `json:"borrowingLimit,omitempty"`
```

If borrowingLimit is nil then cohort must be nil, no bug here. However, we didn't have any validation about this, that's say we can still have cq with `cohort == nil && borrowingLimit != nil`.  

My suggestion here: https://github.com/kubernetes-sigs/kueue/pull/1385#discussion_r1418479788, tl;dr: can we ignore this rule? cc @alculquicondor

Comment (kerthcet) [reply pkg/scheduler/preemption/preemption.go:340]: cc @B1F030

File: pkg/scheduler/preemption/preemption.go [hunk 2]
```diff
@@ -64,7 +64,7 @@ func (p *Preemptor) OverrideApply(f func(context.Context, *kueue.Workload) error
 }
 
 func candidatesOnlyFromQueue(candidates []*workload.Info, clusterQueue string) []*workload.Info {
-	result := make([]*workload.Info, 0)
+	result := make([]*workload.Info, 0, len(candidates))
```
Comment (mimowo) [inline pkg/scheduler/preemption/preemption.go:67]: nit: I would suggest to revert unrelated changes

File: pkg/scheduler/preemption/preemption.go [hunk 3]
```diff
@@ -94,8 +94,8 @@ func (p *Preemptor) GetTargets(wl workload.Info, assignment flavorassigner.Assig
 	// trying to preempt more own workloads and borrow at the same time.
 
 	if len(sameQueueCandidates) == len(candidates) {
-		// There is no risk of preemption of workloads from the other queue,
-		// so we can try borrowing.
+		// There is no need of preemption of workloads from other queues,
```
Comment (mimowo) [inline pkg/scheduler/preemption/preemption.go:97]: nit: I would suggest to revert unrelated changes

File: pkg/scheduler/preemption/preemption.go [hunk 4]
```diff
@@ -174,6 +174,7 @@ func minimalPreemptions(wl *workload.Info, assignment flavorassigner.Assignment,
 			break
 		}
 	}
+
```
Comment (mimowo) [inline pkg/scheduler/preemption/preemption.go:177]: nit: unrelated formatting (and in a couple other places)

**Review by mimowo — COMMENTED** [review]
It is a pity we don't have validation against `cq.Cohort=nil` and `borrowingLimit!=nil`.

Maybe it is still not too late to add, Kueue is still Beta API? 

Anyway, if we go with the check for `cq.Cohort != nil`, should we also add it [here](https://github.com/kubernetes-sigs/kueue/blob/11c46b02cb002364b6db47e84cb97bf392ca8a86/pkg/scheduler/flavorassigner/flavorassigner.go#L550)?

**Comment by kerthcet** [comment]
As you reminded, we're beta now, we can not break the API semantics, so made it a cleanup. And will add the validations about the API. Maybe we can change that at v1beta2 if needed. PTAL @mimowo

File: pkg/scheduler/preemption/preemption.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/preemption/preemption.go:97]:
```suggestion
		// There is no possible preemption of workloads from other queues,
```

File: pkg/scheduler/preemption/preemption.go (hunk 2 above)
Comment (alculquicondor) [reply pkg/scheduler/preemption/preemption.go:67]: also, all the workloads might be from a different CQ.

So it could be better to do `var result []*workload.Info`

Comment (kerthcet) [reply pkg/scheduler/preemption/preemption.go:67]: It's clear here that the length wouldn't greater than len(candidates).

File: pkg/scheduler/preemption/preemption.go (hunk 3 above)
Comment (kerthcet) [reply pkg/scheduler/preemption/preemption.go]: Done.

**Comment by kerthcet** [comment]
Next, I'll add more validations to avoid this.

**Comment by B1F030** [comment]
> Next, I'll add more validations to avoid this.

I write a part of validation in webhook: https://github.com/kubernetes-sigs/kueue/pull/1385/commits/75d6a98ea5da700cb6f58748cc228edb0d084c91

**Review by alculquicondor — COMMENTED** [review]
/approve
/lgtm
